### DISCUSSION TITLE
Why `brew` installs `gcc` and other dev tools as a bottled package dependency?

### DISCUSSION THREAD

**Discussion by abitrolly** [Homebrew/discussions#2713] [Linux]
I am setting up CI and manually install `brew` in a container. When I try to install `duf` utility written in go, `brew` suddenly pulls a lot of packages together with a bottle which contains `duf` binary. How it decides that it needs all these dependencies?

I don't see any dependencies needed for install https://formulae.brew.sh/formula/duf and it doesn't look like `brew` is building anything.

```shell
$ brew install duf
==> Downloading https://ghcr.io/v2/homebrew/core/xz/manifests/5.2.5
==> Downloading https://ghcr.io/v2/homebrew/core/xz/blobs/sha256:c94b4200b32d1e4e917f3b502eafc39579a84533ef6b6a0d58469fa845511612
==> Downloading from https://pkg-containers.githubusercontent.com/ghcr1/blobs/sha256:c94b4200b32d1e4e917f3b502eafc39579a84533ef6b6a0d58469fa845511612?se=2022-01-02T15%3A25%3A00Z&
...
==> Downloading https://ghcr.io/v2/homebrew/core/duf/manifests/0.6.2
==> Downloading https://ghcr.io/v2/homebrew/core/duf/blobs/sha256:272f0ed84e4f0f187f4b07e9200706b872cb364efcfc30e11d832f1f806c42dd
==> Downloading from https://pkg-containers.githubusercontent.com/ghcr1/blobs/sha256:272f0ed84e4f0f187f4b07e9200706b872cb364efcfc30e11d832f1f806c42dd?se=2022-01-02T15%3A25%3A00Z&sig=0SOrcZyPlw0yZ1poMX%2BmBOhNJmyOWJmirEp1LcW5%2FCg%3D&sp=r&spr=https&sr=b&sv=2019-12-12
==> Installing dependencies for duf: xz, m4, gmp, isl@0.18, mpfr, libmpc, zlib, binutils and gcc@5
==> Installing duf dependency: xz
==> Pouring xz--5.2.5.x86_64_linux.bottle.tar.gz
...
 ==> Installing duf dependency: gcc@5
==> Pouring gcc@5--5.5.0_7.x86_64_linux.bottle.tar.gz
Warning: The post-install step did not complete successfully
You can try again using:
  brew postinstall gcc@5
==> Summary
🍺  /home/linuxbrew/.linuxbrew/Cellar/gcc@5/5.5.0_7: 1,348 files, 172.7MB
==> Installing duf
==> Pouring duf--0.6.2.x86_64_linux.bottle.tar.gz
🍺  /home/linuxbrew/.linuxbrew/Cellar/duf/0.6.2: 5 files, 3.2MB
==> Running `brew cleanup duf`...
 Disable this behaviour by setting HOMEBREW_NO_INSTALL_CLEANUP.
Hide these hints with HOMEBREW_NO_ENV_HINTS (see `man brew`).
Error: Process completed with exit code 1.
```

When I run `brew install duf` locally I got no additional dependencies fetches.
```shell
$ brew install duf
==> Downloading https://ghcr.io/v2/homebrew/core/duf/manifests/0.6.2
######################################################################## 100.0%
==> Downloading https://ghcr.io/v2/homebrew/core/duf/blobs/sha256:272f0ed84e4f0f187f4b07e9200706b872cb364efcfc30e11d832f1f806c42dd
==> Downloading from https://pkg-containers.githubusercontent.com/ghcr1/blobs/sha256:272f0ed84e4f0f187f4b07e9200706b872cb364efcfc30e11d832f1f806c42dd?se=2022-01-02T15%3A5
######################################################################## 100.0%
==> Pouring duf--0.6.2.x86_64_linux.bottle.tar.gz
🍺  /home/linuxbrew/.linuxbrew/Cellar/duf/0.6.2: 5 files, 3.2MB
==> Running `brew cleanup duf`...
Disable this behaviour by setting HOMEBREW_NO_INSTALL_CLEANUP.
Hide these hints with HOMEBREW_NO_ENV_HINTS (see `man brew`).                       
````

**Comment by SMillerDev** [comment]
Very likely the gcc on your system is new enough, but the one on CI is not.

**Reply by abitrolly** [reply]
CI is `ubuntu:21.04` docker image, but why `gcc` is needed at all for installation of binary bottle?

**Reply by SMillerDev** [reply]
I believe gcc in homebrew provides the glibc that software links against.

CC @homebrew/Linux for more details

**Reply by abitrolly** [reply]
Does that mean `gcc` is not optional then?

**Reply by carlocab** [reply]
A pre-existing installation of `gcc` is required for `brew` on Linux. See https://docs.brew.sh/Homebrew-on-Linux#requirements. `brew` will install `gcc@5` if your system `gcc` is too old or non-existent. (Though it should probably just error out in the non-existent case, since trying to install `gcc@5` won't work anyway, as you can see in the error shown in your post.)

**Reply by abitrolly** [reply]
Thanks for the explanation. I can understand that `gcc` is needed for compiling packages from source, but why it is needed for bottles?

**Reply by sjackman** [reply]
C++ executables depend on the shared library `libstdc++.so.6` which is provided by GCC. Homebrew checks that the executable `gcc` (or `gcc-*` like `gcc-7`) is available to determine if GCC is installed. If a `gcc` executable is not found, Homebrew installs brewed `gcc` to provide `libstdc++.so.6`. Unfortunately, Homebrew does not know specifically which bottles depend on `libstdc++.so.6`—that's a would-be-nice-to-fix issue in the Homebrew dependency system—so it assumes that all bottles depend on `libstdc++.so.6`.

**Reply by abitrolly** [reply]
Can `ldd` help with detection if the binary depends on some shared lib?
```shell
$ ldd  $(readlink -f `which duf`)
	not a dynamic executable
$ echo $?
1
```
Another question why install `gcc` instead of just [libstdc++6](https://packages.ubuntu.com/hirsute/libstdc++6)?

**Reply by SMillerDev** [reply]
Because the latter is not a homebrew package.
